pkg/controller/endpointslice/reconciler.go

```diff
@@ -56,12 +57,62 @@ type endpointMeta struct {
 // slices for the given service. It creates, updates, or deletes endpoint slices
 // to ensure the desired set of pods are represented by endpoint slices.
 func (r *reconciler) reconcile(service *corev1.Service, pods []*corev1.Pod, existingSlices []*discovery.EndpointSlice, triggerTime time.Time) error {
-	addressType := discovery.AddressTypeIPv4
+	slicesToDelete := []*discovery.EndpointSlice{}                                    // slices that are no longer  matching any address the service has
+	errs := []error{}                                                                 // all errors generated in the process of reconciling
+	slicesByAddressType := make(map[discovery.AddressType][]*discovery.EndpointSlice) // slices by address type
 
-	if endpointutil.IsIPv6Service(service) {
-		addressType = discovery.AddressTypeIPv6
+	// addresses that this service supports [o(1) find]
+	serviceSupportedAddresses := getAddressTypesForService(service)
+
+	// loop through slices identifying their address type.
+	// slices that no longer match address supported by services
+	// go to delete, other slices goes to the reconciler machinery
+	// for further adjustment
+	for _, existingSlice := range existingSlices {
+		// service no longer support that address type, add it to deleted slices
+		if _, ok := serviceSupportedAddresses[existingSlice.AddressType]; !ok {
+			slicesToDelete = append(slicesToDelete, existingSlice)
+			continue
+		}
+
+		// add list if it is not on our map
+		if _, ok := slicesByAddressType[existingSlice.AddressType]; !ok {
+			slicesByAddressType[existingSlice.AddressType] = make([]*discovery.EndpointSlice, 0)
```

Comment:
why not set capacity to 1 - you know you will use the 1 below and most services have just 1 slice...